### PR TITLE
Slightly better webpack desktop 1918

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

Makes webpack ~5x faster on my machine.

Moves all our vendored code to something called a DLL which is build as a separate step. Then only our app code is watched by the webpack dev server.

For me I see incremental build times take between 1 - 5 s.

### Future Improvements

Maybe using happy pack could get us even faster, I haven't tried that.

There are a couple new npm commands:

`npm run hot-server` should work just as before but have a slower start, faster incremental builds.
This is because it builds the main thread bundle, then the dll bundle, then starts the hot server.

`npm run just-hot-server` only does the hot server instead of the three steps above

`npm run build-dll` builds the dll bundle

`npm run build-main-thread` builds the main thread bundle

`npm run build-profile` profiles the dev webpack build and puts it in `/tmp/stats.json`

### CODE REVIEW

File: desktop/webpack.config.development.js [hunk 1]
```diff
@@ -18,7 +19,8 @@ const defines = {
 console.warn('Injecting dev defines: ', defines)
 
 config.debug = true
-config.devtool = NO_SOURCE_MAPS ? undefined : 'inline-source-map'
+config.cache = true
+config.devtool = NO_SOURCE_MAPS ? undefined : 'eval-source-map'
```
Comment (chromakode) [inline desktop/webpack.config.development.js]: Did you intend to change this again? I seem to recall we just changed it due to issues with `eval-source-map` in Chrome.

File: desktop/webpack.config.development.js [hunk 2]
```diff
@@ -73,5 +84,11 @@ if (getenv.boolish('HOT', false)) {
     }
   })
 }
+
+if (USING_DLL) {
+  // Don't build the main thing
```
Comment (chromakode) [inline desktop/webpack.config.development.js]: This comment is a bit terse. Could you please expand upon it to clarify why we're not building it?

File: desktop/renderer/index.js [hunk 3]
```diff
@@ -23,7 +23,7 @@ import {setupContextMenu} from '../app/menu-helper'
 import electron, {ipcRenderer} from 'electron'
 import {ipcLogsRenderer} from '../app/console-helper'
 import loadPerf from '../shared/util/load-perf'
-import merge from 'lodash/merge'
+import {merge} from 'lodash'
```
Comment (chromakode) [inline desktop/renderer/index.js]: Ah I see, only `lodash` is vendored.

File: desktop/webpack.config.dll-build.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+var path = require('path')
+var webpack = require('webpack')
+
+module.exports = {
+  entry: {
+    vendor: ['material-ui', 'react', 'lodash', 'immutable', 'material-ui/styles', 'material-ui/svg-icons', 'material-ui/FlatButton', 'material-ui/Popover', 'moment', 'lodash', 'html-entities', 'redux-saga', 'core-js', 'qrcode-generator', 'react-json-tree', 'redux'],
```
Comment (chromakode) [inline desktop/webpack.config.dll-build.js]: `lodash` is listed twice in this list. Perhaps sort it to make it easier to catch dupes?

File: shared/profile/dumb.js [hunk 5]
```diff
@@ -170,9 +170,9 @@ const propsBase: RenderProps = {
       height: 578,
     },
   },
-  onClickAvatar: console.log('on click avatar'),
-  onClickFollowers: console.log('on click followers'),
-  onClickFollowing: console.log('on click following'),
+  onClickAvatar: () => console.log('on click avatar'),
```
Comment (chromakode) [inline shared/profile/dumb.js]: Thanks for fixing this. Wasn't expecting to see it in this commit, though. Separate PR next time please?

**Comment by keybase-ci-visdiff** [comment]
The commits 0229d49b1d97...5841d55f1a10 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **unlock_folders-Normal** [(before)](https://keybase-app-visdiff.s3.amazonaws.com/1475539533156-0229d49b1d97-5841d55f1a10-linux/unlock_folders-Normal-0229d49b1d97.png) [(after)](https://keybase-app-visdiff.s3.amazonaws.com/1475539533156-0229d49b1d97-5841d55f1a10-linux/unlock_folders-Normal-5841d55f1a10.png)  
   unlock_folders-Normal comparison
</summary>

File: desktop/webpack.config.development.js (hunk 1 above)
Comment (MarcoPolo) [inline desktop/webpack.config.development.js]: worked for me, we can change if it causes problems

File: desktop/client.js [hunk 6]
```diff
@@ -8,32 +7,12 @@ try {
...
-  req.on('error', e => {
-    console.log('Error: ', e)
-    console.log('Sleeping 5 seconds and retrying. Maybe start the hot-server?')
-    setTimeout(() => hitServer(), 5000)
-  })
+if (process.env.USE_INSPECTOR) {
+  params.unshift('--debug-brk=5858')
 }
 
-hitServer()
```
Comment (chrisnojima) [inline desktop/client.js]: does this still function the same as the previous behavior?

**Comment by gabriel** [comment]
Can we rename `just-hot-server` to `hot-server-only` so the commands appear in the list next to each other.

**Comment by gabriel** [comment]
Should we rename use of "dll"? It just screams windows to me and could be confusing.

**Comment by MarcoPolo** [comment]
Jenkins says green on the interface so I'm merging

**Pull request merged**